dense_vector/track.py

```diff
@@ -1,5 +1,110 @@
+        self._store = defaultdict(lambda: defaultdict(list))
+
+    async def get_neighbors_for_query(self, index: str, query_id: str, max_size: int, client) -> List[str]:
+        try:
+            logger.debug(f"Fetching exact neighbors for {query_id} from in-memory store")
+            if not (index in self._store and query_id in self._store[index]):
+                logger.debug(f"Query vector with id {query_id} not cached - computing neighbors")
+                self._store[index][query_id] = await self.load_exact_neighbors(index, query_id, max_size, client)
+                logger.debug(f"Finished computing exact neighbors for {query_id} - it's now cached!")
+            return self._store[index][query_id]
+        except Exception as ex:
+            logger.error(f"Failed to compute nearest neighbors for '{query_id}'. Returning empty results instead.", ex)
+            return []
+
+    async def load_exact_neighbors(self, index: str, query_id: str, max_size: int, client):
+        if query_id not in self._query_vectors:
+            raise ValueError(f"Unknown query with id: '{query_id}' provided")
+        return await extract_exact_neighbors(self._query_vectors[query_id], index, max_size, client)
+
+    def get_query_vectors(self) -> Dict[str, List[float]]:
+        return self._query_vectors
+
+    @classmethod
+    @functools.lru_cache(maxsize=1)
+    def get_instance(cls, queries_file: str):
+        logger.info(f"Initializing KnnVectorStore for queries file: '{queries_file}'")
+        return KnnVectorStore(queries_file)
+
+
+class KnnRecallProcessor:
+    KNN_RECALL_OPERATION_PATTERN = re.compile(r"knn-recall-(\d+)-(\d+)")
+
+    def on_after_load_track(self, t):
+        max_k = 0
+        tasks_to_update = []
+        for challenge in t.challenges:
+            for task in challenge.schedule:
+                if isinstance(task, Task):
+                    task_matched = self.KNN_RECALL_OPERATION_PATTERN.search(task.operation.name)
+                    if task_matched:
+                        tasks_to_update.append(task)
+                        matched_k_parameter = task_matched.group(1)
+                        max_k = max(max_k, int(matched_k_parameter))
+                elif isinstance(task, Parallel):
+                    for nested_task in task.tasks:
+                        task_matched = self.KNN_RECALL_OPERATION_PATTERN.search(nested_task.operation.name)
+                        if task_matched:
+                            tasks_to_update.append(nested_task)
+                            matched_k_parameter = task_matched.group(1)
+                            max_k = max(max_k, int(matched_k_parameter))
+            for task in tasks_to_update:
+                task.operation.params.update({"max_k": max_k})
+
+    def on_prepare_track(self, track, data_root_dir):
+        return []
+
+    def __repr__(self, *args, **kwargs):
+        return "knn-recall-processor"
```

Comment:
I think we can assume that `max_k=1000` or even `max_k = 10000` and simplify the code (get rid of `KnnRecallProcessor`) , what do you think of this?

`script_score` query performance doesn't change much depending on `size` as we need to go through all vectors regardless.

Reply:
Yeah I see your point! +1 for `max_k=1_000` (`10_000` might be an overkill for 99% of the cases and we would just end up storing potentially thousands of these lists in memory). We can default to `1_000`, and if an operation with more results comes in, we can make an overrequest for that specific task.

Reply:
Updated.